### ISSUE TITLE
Issue with possible loop on run

### ISSUE THREAD

**Issue by jwoertink** [jgaskins/redis#12]
I'm trying to integrate this shard in to Cable using the examples by @mjeffrey18 https://github.com/jgaskins/redis/issues/7#issuecomment-1166516550 but when I run the specs for Cable they seem to just hang after the first spec.

My guess is since we call `Cable.restart` between each spec, that runs the `shutdown` which calls this:

```crystal
def shutdown
  redis_subscribe.run({"unsubscribe"})

  redis_subscribe.close
  redis_publish.close
  connections.each do |_, v|
    v.close
  end
end
```

If I add some puts after that `redis_subscribe.run`, it never prints. Now, if I swap that out for this:

```crystal
redis_subscribe.unsubscribe("")
```

That will run the specs, though, I get a bunch of `IO::Error`

**Comment by jgaskins** [comment]
Looks like `UNSUBSCRIBE` doesn't return anything, but `run` expects to read a response from the server. So `Connection#encode` plus a `flush` should do the trick. I just had a look and it turns out that's actually the exact implementation of [`Connection#unsubscribe`](https://github.com/jgaskins/redis/blob/10138fd6ad98d5751c882632b3fd15f5010fa189/src/connection.cr#L141-L144). Looks like Cable also took that into account originally by only [writing to the server](https://github.com/cable-cr/cable/pull/48/files#diff-dfd77a42658185844ac9bf5040bb85cb2186824dd5fec0f16686ae69ef578b0bL130-L132) (calling [a very similar method](https://github.com/stefanwille/crystal-redis/blob/fa7121fb3acc4b9ffb951ecb716f767f59b4ce8d/src/redis/connection.cr#L49-L52)) but not reading back from it.

As for the `IO::Error`, that seems weird. It doesn't look like you're reusing the same `Connection` instance because you're using [a fresh `Cable::Server`](https://github.com/cable-cr/cable/blob/a6672c4f19dc25a61b35c403db4b883ae86f834e/src/cable/server.cr#L11-L16), which should give you a fresh `Redis::Connection`, which should give you a fresh `TCPSocket`.

**Comment by jgaskins** [comment]
I just noticed why you were trying to use `run`. It didn't occur to me that splat args in Crystal requires at least one arg (a lesson I learn every couple months and promptly forget), so there's no way to send a plain `unsubscribe` with the current implementation.

**Comment by jwoertink** [comment]
Yeah, I want to unsubscribe from all channels, so I have to at least pass 1 arg. It sounds like this is the right path to just use `unsubscribe()` instead of `run()` here. Thanks for the insight.

I'm still not sure why the IO::Error on the specs, so I'll keep digging around on that. I must be missing something.

**Comment by jgaskins** [comment]
Cool, I'll add some method overloads to accommodate that.

I've been playing with this, too, and I can't seem to reproduce it without Cable. I'll post my findings on https://github.com/cable-cr/cable/pull/48.

**Comment by jgaskins** [comment]
Oh wait, I think I know what this is. It's not happening when the new connection spins up, which is what I thought was happening. It's that the `Redis::Parser#read` from your _previous_ spec is still running in the other fiber when the connection is closed. You're correct, the problem is that this shard is not accounting for the connection closing.

I was able to reproduce *that* behavior very simply:

```crystal
require "../src/redis"

subscriber = Redis::Connection.new

spawn { subscriber.subscribe "foo" {} }

subscriber.close
```

I'll try to have a fix out (along with the empty `unsubscribe` overload) this evening.

**Comment by jgaskins** [comment]
This also explains why the exception wasn't failing your specs.

**Comment by jwoertink** [comment]
Nice! Glad you were able to find it. Thanks for the help on it.

**Comment by jgaskins** [comment]
Released v0.5.0 with the fix for this. I confirmed that the specs on https://github.com/cable-cr/cable/pull/48 pass locally with this change.

**Comment by jwoertink** [comment]
Sweet! I ran it locally and most of those IO error issues did go away. There does still seem to be 1 left, but maybe not related?

```
cable on  issues/46 is 📦 v0.2.1 via 🔮 v1.4.1 
❯ crystal spec 
.....Unhandled exception in spawn(name: Cable::Server - subscribe): Closed stream (IO::Error)
  from /usr/share/crystal/src/io.cr:118:5 in 'check_open'
  from /usr/share/crystal/src/io/buffered.cr:165:5 in 'write_byte'
  from /usr/share/crystal/src/char.cr:897:9 in 'to_s'
  from /usr/share/crystal/src/io.cr:174:5 in '<<'
  from lib/redis/src/writer.cr:12:7 in 'encode'
  from lib/redis/src/connection.cr:125:5 in '->'
  from /usr/share/crystal/src/fiber.cr:146:11 in 'run'
  from /usr/share/crystal/src/fiber.cr:98:34 in '->'
  from ???
.....*...........................

```

Digging in to this one, it looks like the handler spec calls `Cable.restart` a second time..
https://github.com/cable-cr/cable/blob/a6672c4f19dc25a61b35c403db4b883ae86f834e/spec/cable/handler_spec.cr#L107

Removing this line fixes (which is nice), but it seems calling that `Cable.restart` twice in a row borks it... I'm gonna just update the spec and move on though, this is awesome!

**Comment by jgaskins** [comment]
I did see that once when I ran it, but it didn't show up again. I think it's due to asynchrony in `spawn` in [`Cable::Server#subscribe_to_internal`](https://github.com/cable-cr/cable/blob/1366d8c93b8f959461d67fc5098bb5b38a76ba20/src/cable/server.cr#L131-L143). By the time that fiber runs for the first `restart`, the second one may or may not have closed its socket. Increased latency to the server might make it more reliably reproducible.

If I understand the code in Stefan's client (and there is no guarantee for that), it may have been silently reconnecting to send the `SUBSCRIBE` command instead of raising an exception:

- `subscribe` [calls `void_command`](https://github.com/stefanwille/crystal-redis/blob/fa7121fb3acc4b9ffb951ecb716f767f59b4ce8d/src/redis/commands.cr#L1741)
- `void_command` [calls `command`](https://github.com/stefanwille/crystal-redis/blob/fa7121fb3acc4b9ffb951ecb716f767f59b4ce8d/src/redis/command_execution/value_oriented.cr#L88)
- `command` [calls `with_reconnect`](https://github.com/stefanwille/crystal-redis/blob/fa7121fb3acc4b9ffb951ecb716f767f59b4ce8d/src/redis.cr#L310-L312)

This implementation doesn't do that because, while [`Connection#run` _does_ reconnect and retry if a connection is lost](https://github.com/jgaskins/redis/blob/7f9ca239e0d635ec86f51717960750a5fb7c911b/src/connection.cr#L241-L254), it can be dangerous to do that for `encode`, since it's used by things like pipelines. For example, if we disconnect in the middle of a pipeline block, we may try to read the results of 5 commands but the server only sends 3 results. Coincidentally, I added [a warning to the docs](https://jgaskins.dev/redis/Redis/Connection.html#encode%28command%29-instance-method) about this right as you posted this comment.

**Comment by jwoertink** [comment]
Ah, that makes sense. Good call. Yeah, I think calling that twice in a row should be avoided anyway, so definitely not an issue in this shard. :+1:
